File: kotlin/src/main/kotlin/de/florian/adventofcode/y2020/Day06.kt
package de.florian.adventofcode.y2020

import de.florian.adventofcode.AoCDay

fun main() { Day06().exec() }

class Day06 : AoCDay() {
    private fun convertInput(input: String): List<List<MutableSet<Char>>> = input.split(Regex("\\n\\n")).map { group ->
        group.trim().split("\n").map { it.toSortedSet()}
    }

    private val answers = convertInput(Inputs_2020.DAY_06)

    override fun part1(): String {
        return answers.map {
            it.fold<Set<Char>, Set<Char>>(mutableSetOf(), { acc, e -> (acc + e) }).size
        }.sum().toString()
    }

    override fun part2(): String {
        return answers.map { group ->
            val map = mutableMapOf<Char, Int>()
            group.forEach { personAnswers ->
                personAnswers.forEach { answer ->
                    map.merge(answer, 1){a,b -> a+b}
                }
            }
            map.filter { it.value == group.count() }.size
        }.sum().toString()
    }
}

File: typescript/src/2022/day06.ts
import {AocDay, AocDayDecorator, SolutionParts} from "../aoc";

@AocDayDecorator('2022', '6')
export class Day06 extends AocDay {

	solveImpl(): SolutionParts {
		const findIdxStartOfPacketMarker = (sequenceLen: number): number => {
			const chars = this.input.trim().split("");
			for (let i = 0; i < chars.length; i++)
				if(new Set(chars.slice(i, i+sequenceLen)).size >= sequenceLen)
					return i + sequenceLen;
			return -1;
		};

		return {
			part1: findIdxStartOfPacketMarker(4),
			part2: findIdxStartOfPacketMarker(14),
		};
	}
}